juju/juju
# [JUJU-4383] Defer removal of application secrets until the app is removed, not just destroyed

When an app is destroyed, secrets were being removed as soon as the app transitioned to Dying. This meant that the secrets could not be read from any relation departed, storage detaching hooks etc. Now the secrets are removed only after the app is removed.

Also, update upstream juju/charm to include secret hooks in debug-hooks.

## QA steps

I created a charm with code to access the charm's secret(s) in a relation departed hook.
Deploy the charm and relate it; create a secret.
Destroy the app and inspect the logs to see that the charm can read the secret in the relation departed hook.

Deploy charm and relate it; add a secret and grant access.
run debug-hooks on the secret-changed hook for the consuming unit.

Drive by - update go.mod to remove some CVEs.

## Bug reference

https://bugs.launchpad.net/juju/+bug/2029285
https://bugs.launchpad.net/juju/+bug/2029282
